keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,494 @@
+* Selected API should support structured logging by providing a way to define a set of key-value pairs for each log entry.
+* There should at least one production ready implementation of the selected API which could produce JSON formatted logs.
+* It should be possible to produce klog compatible logs from the selected API which could be used during the migration to a new API.
+
+Based on this criteria https://github.com/go-logr/logr log interface has been selected as a logging API for Kubernetes.
+
+Apart from meeting the criteria listed above logr has a few nice features:
+* Constant log message together with logger name provide stable identification for log entries which greatly simplifies querying logs.
+* Limiting log levels to just Info and Error makes it much easier to choose the proper log level for given log.
+* Using key-values pairs for defining log fields allows more efficient implementation of the logger.
+
+#### Logger initialization
+
+Logger implementation initialization should be done at startup of component and passed down the stack as an interface. Aside of initialization of logger, there should not be any dependency on specific implementation of logger Interface. 
+
+#### Storing Logger and metadata in Context
+
+We would like to propose that logger interface and all metadata collected trough call stack to be stored in Golang Context object. This way creating one canonical way to pass contextual information across the call stack. Some pros of this approach:
+* Standard method in Golang to pass scoped variables across API boundaries and between processes.
+* No need for additional function argument for passing logging API through call stack
+* Context is used similarly by many frameworks including for instance OpenCensus tracing API.
+* Accessing metadata will be independent from Logging and could be reused by other functionalities.
+
+Access to context variables should be done though predefined functions. For that we propose to a new module to kubernetes component base called `kontext`. This new module will include all utility functions for storing and retrieving logger and metadata needed by it. 
+
+Throughout call stack new metadata can be added to context that can be later used by logger. For example reconciliation loop can reference the reconcilided object so that all logs will be referencing it. All metadata access should be done through functions in `kontext` module. Adding reference to kubernetes object (pod, service, node) should be really straightforward.
+
+#### Example
+
+```golang
+func UpdateStatus(ctx context.Context, pods []*v1.Pod, newStatus bool) {
+	for _, pod := range pods {
+		podCtx := kontext.WithObject(ctx, pod)
+		updatePodStatus(podCtx, pod, newStatus)
+	}
+}
+
+func updatePodStatus(ctx context.Context, pod *v1.Pod, newStatus bool) {
+  logger := kontext.GetLogger(ctx)
```

Comment:
What happens when `ctx` doesn't have a `logr` instance? Is that an error?

The problem then will be that introducing logging via the context is a very subtle API change because passing `context.Background()` was okay so far and suddenly isn't sufficient anymore, without triggering compiler errors.

Reply:
Good question. I see at least a few solutions for this problem:
- return error - not very practical as the caller will have to either ignore error silently which is usually a bad thing or to propagate it which will break the execution.
- return no-op logger - this is basically the same solution as returning an error and always ignoring it which is far from being perfect as it will make ignored logs invisible to users.
- return default logger - probably the easiest to implement and the most robust solution. By setting some specific name to the returned logger allows tracking logs without a proper context. If the default logger is not set it could do `os.Exit(...)`. 
- replace `context.Context` with custom `kontext.Context` interface - allows to introduce a stricter contract for a new interface which for instance enforces that some Logger must always be attached to the context and allows compiler to report places where we pass `context.Background()` as a context.  This however might not not be completely bulletproof for instance in case of some external libraries which already use `context.Context` and call some callbacks because for those we might need to go through `kontext.Context` -> `context.Context` -> `kontext.Context` casting.

Out of those options the third one seems for me to be the best.

Reply:
> return default logger - probably the easiest to implement and the most robust solution. By setting some specific name to the returned logger allows tracking logs without a proper context.

That's the solution that I also arrived at when thinking about this a while back. The problem then becomes that we either need to deal with no default being set or tying a mostly abstract module like `logr` to a specific, simplistic implementation (just print to stderr, for example) that is the global default when not being overridden.

>  If the default logger is not set it could do os.Exit(...).

I don't like that personally. When using some third-party library that happens to implement logging via this new API, then I'd rather take some spurious output in some code path rather than my whole app dying because of some info message.

Reply:
Calling `os.Exit(...)` when default global logger is not set should not be a problem for normal execution as default logger should be set while initializing logr implementation. So in normal circumstances default logger will always be set. The only case when this might not be the case are in my option tests.

Reply:
This is only true under two assumptions:
1. the application knows that it needs to initialize a logr implementation and does so
2. no logging calls are triggered before that initialization happens

I'm kind of okay with the second one, because a package which triggers log messages during its init phase is suspicious; it probably shouldn't do that.

But the first one is a problem. Suppose an application is importing some utility package and that package starts to use logr instead of, say, printing to stderr, then that's a hidden API change of that package because the application must be changed to provide a logr implementation.

The utility package could have a fallback logger, but it shouldn't set that as global default (because merely importing a package shouldn't affect global defaults) and then we are back at the case where each `kontext.GetLogger(ctx)` call must be checked for nil. An application could also end up with multiple different fallbacks if its packages don't agree on one.

Reply:
Just rephrasing what have been said before to make sure that we are on the same page:

With logr being just an abstract logging API third party libraries which are using it should not need to care about the actual implementation of this API. They should simply use it and it should be the responsibility of the caller which uses given library to provide some implementation of this API.
The way to provide this implementation could be either explicit by for instance simply passing an instance of `logr.Logger` as an argument to some library function or more implicit by expecting that it will be associated with `context.Context` which is passed to the library.
In the second case, as we have discussed above, there is no easy way to make caller aware that the logger should be passed and force caller to so without changing `context.Context` to a more specialized interface.
To address this problem, as we have agreed, it would be good to have a global default fallback logger instance which could be used when there is no logger within the context. This default logger should be initialized by the application while the application is initialized.
The main problem with this approach is the fact that application needs to know about it and this might be a problematic if application by itself does not use logr and it is just some dependency which does. In such case the need of initializing default logger might go unnoticed.
In such case in my option the best thing that we can do is to detect this situation as soon as possible and let developer or user know about it either by writing some error message on stderr or by stopping application. In would be optimal if this detection could be done during the application compilation or application startup but I do not know any good way to do so.
So the best solution which I can offer right now is to do this check in `kontext.GetLogger(ctx)` and suggest to third party libraries developers that switch to logr should be considered a breaking API change and a new major version of library should be released after such change.

Reply:
I think we are on the same page, but I come to a different conclusion. A breaking API change that is just a comment in the release note saying "you must now initialize a global logr instance" will be overlooked. If the logging then happens in an untested error path, "bad things will happen" (TM).

I prefer to always have a global logger as fallback. I don't care too much whether it is a null logger (discards log messages) or something that writes to stdout with minimalistic formatting. I'm leaning a bit more towards the latter because then developers or users will notice that logging is supported and could be enhanced by choosing a better implementation and because log messages might be needed for debugging.

Reply:
In such case there is always an option for the library to include some logr implementation as a dependency and initialize it as a default logger in its init method if this has not been set yet.
This way we will end up with 3 libraries which need to be added as dependencies:
- go-logr/logr
- go-logr/context
- go-logr/stdr

```go
func init() {
    if lcontext.getDefaultLogger() == nil {
        lcontext.setDefaultLogger(stdr.New(log.New(os.Stderr, "", log.LstdFlags))))
    }
}
```

Reply:
It looks wrong to me to have a dependency from a library to a specific logr implementation. Isn't that what we are trying to get away from?

Reply:
Furthermore, if that is how we expect libraries to use `go-logr/context`, why not put that into `go-logr/context` itself?

Reply:
Yes, having a dependency on specific logr implementation is definitely wrong. But moving an init function and the dependency to `go-logr/context` or some other library in my opinion does not really solve the problem as the implementation is still there just hidden. For new libraries or libraries with new major versions of libraries this should not be needed. Therefore I would prefer not to include this default implementation in `go-logr/context` directly. But wrapping it with some other library might be indeed convenient. In such case imports could look like this:
```go
import (
  "github.com/go-logr/logr"
  lcontext "github.com/go-logr/context"
  _ "github.com/go-logr/default-nop-logger"
)
```

Reply:
Thanks for your feedback. Context-ful logging and storing logger on context is a topic that requires dedicated discussion. We decided to remove it from scope of this KEP